stdlib/condition.mli

```diff
@@ -13,41 +13,112 @@
+   side, one or more threads can indicate that they wish to {i wait}
+   for a certain property to become true. On the sender side, a thread
+   can {i signal} that this property has become true, causing one (or
+   more) waiting threads to be woken up.
+
+   For instance, in the implementation of a queue data structure, if a
+   thread that wishes to extract an element finds that the queue is
+   currently empty, then this thread waits for the queue to become
+   nonempty. A thread that inserts an element into the queue signals
+   that the queue is nonempty. A condition variable is used for this
+   purpose. This communication channel conveys the information that
+   the property "the queue is nonempty" is true, or (more accurately)
+   may be true.
+
+   In short, a condition variable [c] is used to convey the information
+   that a certain property [P] about a shared data structure [D],
+   protected by a mutex [m], may be true.
+
+   With a condition variable [c], exactly one mutex [m] is associated.
+   This association is implicit: the mutex [m] is not explicitly passed
+   as an argument to {!create}. It is up to the programmer to know, for
+   each condition variable [c], which is the associated mutex [m].
+
+   With a mutex [m], several condition variables can be associated.
+   For instance, in a bounded queue, one condition variable could
+   be used to indicate that the queue is nonempty, and another condition
+   variable could be used to indicate that the queue is not full.
+
+   With a condition variable [c], exactly one logical property [P]
+   should be associated. Examples of such properties
+   include "the queue is nonempty" and "the queue is not full".
+   It is up to the programmer to keep track, for each condition
+   variable, of the corresponding property [P].
+   A signal should be sent on the condition variable [c]
+   only when the property [P] is definitely true.
+   When a signal is received, however, one cannot assume that [P] is
+   definitely true; one must explicitly test whether [P] is true.
+   The reason is that,
+   between the moment when the signal is sent
+   and the moment when a waiting thread receives the signal
+   and is scheduled,
+   the property [P] may be falsified by some other thread
+   that is able to acquire the mutex [m] and alter the data structure [D].
+
+   Assuming that [D] is a shared data structure
+   protected by the mutex [m],
+   and assuming that [c] is a condition variable
+   associated with the mutex [m]
+   and with the property [P],
+   a typical usage scenario is as follows:
    {[
      Mutex.lock m;
-     while (* some predicate P over D is not satisfied *) do
+     while (* the property P over D is not satisfied *) do
        Condition.wait c m
      done;
      (* Modify D *)
-     if (* the predicate P over D is now satisfied *) then Condition.signal c;
+     if (* the property P over D is now satisfied *) then Condition.signal c;
```

Comment:
I'm not sure this is a good example...  In typical examples like bounded buffers, waiting and signaling are done on different conditions, and signaling is not guarded by an "if":
```
  while not (is_empty q) do Condition.wait q.nonempty m done;
  let x = take q in
  Condition.signal q.notfull;
  x
```
I can't think of a good example that follows the "one condition" pattern given in the docs.

Why not show the bounded buffer example, while we're at it?

Reply:
Indeed, the code snippet (which I did not change) is weird, as it tries to illustrate both sides (waiting and signaling) using a single code fragment and a single condition variable. I will change it to show two separate snippets taken from the bounded buffer example.